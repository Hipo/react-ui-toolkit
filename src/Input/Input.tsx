import React from "react";
import classNames from "classnames";
import styled from "styled-components";

const InputContainer = styled.div`
  display: flex;
`;
const PrefixIcon = styled.span`
  background: white;

  border-right: 1px solid grey;
`;
const SuffixIcon = styled.span`
  background: white;

  border-left: 1px solid grey;
`;

interface InputProps {
  name: string;
  type?: string;
  value?: string;
  onChange: React.ReactEventHandler<HTMLInputElement>;
  onFocus?: React.ReactEventHandler<HTMLInputElement>;
  onBlur?: React.ReactEventHandler<HTMLInputElement>;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  onKeyUp?: React.KeyboardEventHandler<HTMLInputElement>;
  placeholder?: string;
  min?: number;
  max?: number;
  step?: number;
  role?: string;
  prefixIcon?: React.ReactNode;
  suffixIcon?: React.ReactNode;
  isDisabled?: boolean;
  className?: string;
  inputContainerRef?: React.RefObject<HTMLDivElement>;
}

function Input(props: InputProps) {
  const {
    name,
    type = "text",
    value,
    onChange,
    placeholder,
    step,
    min,
    max,
    isDisabled,
    onFocus,
    className,
    onBlur,
    onKeyUp,
    onKeyDown,
    prefixIcon,
    suffixIcon,
    role,
    inputContainerRef,
    ...rest
  } = props;
  const inputContainerClassName = classNames("input-container", className);
  const inputClassName = classNames("input", {
    disabled: isDisabled
  });

  return (
    <InputContainer
      ref={inputContainerRef}
      role={role}
      className={inputContainerClassName}>
      {prefixIcon && (
        <PrefixIcon className={"input-container-prefix-icon"}>{prefixIcon}</PrefixIcon>
      )}

      <input
        className={inputClassName}
        name={name}
        type={type}
        step={step}
        max={max}
        min={min}
        value={value}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
        placeholder={placeholder}
        autoComplete={"off"}
        autoCorrect={"off"}
        disabled={isDisabled}
        onKeyDown={onKeyDown}
        onKeyUp={onKeyUp}
        {...rest}
      />

      {suffixIcon && (
        <SuffixIcon className={"input-container-suffix-icon"}>{suffixIcon}</SuffixIcon>
      )}
    </InputContainer>
  );
}

export default Input;
