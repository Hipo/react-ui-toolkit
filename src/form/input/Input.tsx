import "./_input.scss";

import React, {useState, useEffect} from "react";
import classNames from "classnames";

import {
  getNumberSeparators,
  parseNumber,
  mapDigitsToLocalVersion,
  formatNumber,
  removeLeadingZeros,
  getNegativeZero,
  getThousandthSeparatorCount
} from "../../core/utils/number/numberUtils";
import {getInputLocalizationOptions, getInputParseNumberOptions} from "./util/inputUtils";
import {InputProps} from "./util/inputTypes";

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  /* eslint-disable complexity */
  (props, ref) => {
    const {
      testid,
      value,
      type = "text",
      isDisabled,
      hasError,
      customClassName,
      leftIcon,
      rightIcon,
      localizationOptions = {},
      role,
      autoComplete = "off",
      autoCorrect = "off",
      onChange,
      ...rest
    } = props;
    const {
      shouldFormatToLocaleString = false,
      locale,
      maximumFractionDigits = 0
    } = getInputLocalizationOptions(localizationOptions);
    const [
      {
        DECIMAL_NUMBER_SEPARATOR: decimalSeparatorForLocale,
        MINUS_SIGN: minusSignForLocale,
        LOCALE_NEGATIVE_ZERO: negativeZeroForLocale
      },
      setNumberSeparatorsForLocale
    ] = useState(() => ({
      ...getNumberSeparators(locale),
      ...getNegativeZero(locale)
    }));
    const isNumberInput = type === "number";
    const inputContainerClassName = classNames(
      "input-container",
      customClassName,
      `input-container--type-${type}`
    );
    const inputClassName = classNames("input", {
      "input-container--is-disabled": isDisabled,
      "input-container--has-error": hasError
    });
    let finalValue = value;

    if (
      !(
        typeof maximumFractionDigits === "number" &&
        Number.isInteger(maximumFractionDigits) &&
        maximumFractionDigits >= 0
      )
    ) {
      throw new Error("maximumFractionDigits should be zero or a positive integer.");
    }

    if (isNumberInput && typeof value === "string" && shouldFormatToLocaleString) {
      const [integerPart, decimalPart] = String(value).split(".");
      const numberFormatter = formatNumber({
        maximumFractionDigits,
        locale
      });

      // IF there is a decimal part or the value ends with ".",
      // make sure we add the decimal separator and map each digit on the decimal part to localized versions.
      // We shouldn't use parseInt or parseFloat with numberFormat util here because that removes zeros on the decimal part and disallows users to write something like: 10.01 or 10.102
      if (value.match(/\.$/)?.length || decimalPart) {
        finalValue = `${numberFormatter(
          parseInt(integerPart)
        )}${decimalSeparatorForLocale}${mapDigitsToLocalVersion({locale}, decimalPart)}`;
      } else if (integerPart) {
        if (integerPart !== minusSignForLocale && integerPart !== negativeZeroForLocale) {
          finalValue = numberFormatter(parseInt(integerPart));
        } else {
          finalValue = `${minusSignForLocale}${mapDigitsToLocalVersion(
            {locale},
            integerPart
          )}`;
        }
      }
    }

    useEffect(() => {
      setNumberSeparatorsForLocale({
        ...getNumberSeparators(locale),
        ...getNegativeZero(locale)
      });
    }, [locale]);

    return (
      <div role={role} className={inputContainerClassName} data-testid={testid}>
        {leftIcon && (
          <span className={"input-container__icon input-container__left-icon"}>
            {leftIcon}
          </span>
        )}

        <input
          ref={ref}
          className={inputClassName}
          type={isNumberInput ? "text" : type}
          autoComplete={autoComplete}
          value={finalValue}
          autoCorrect={autoCorrect}
          disabled={isDisabled}
          onChange={handleChange}
          {...rest}
        />

        {rightIcon && (
          <span className={"input-container__icon input-container__right-icon"}>
            {rightIcon}
          </span>
        )}
      </div>
    );

    function handleChange(event: React.SyntheticEvent<HTMLInputElement>) {
      if (isNumberInput) {
        const {value: newValue} = event.currentTarget;

        if (newValue) {
          const formattedNewValue = parseNumber(
            getInputParseNumberOptions({locale, maximumFractionDigits}),
            newValue
          );
          // Number("-") returns NaN. Should allow minus sign as first character.
          const isFormattedNewValueNotAValidNumber =
            formattedNewValue === "" ||
            (newValue !== "-" && Number.isNaN(Number(formattedNewValue)));
          let finalEventValue = formattedNewValue ? String(formattedNewValue) : "";

          // IF the parsed number is valid and there is a decimal separator,
          // we need to save the number as it is so that decimal part doesn't disappear
          if (
            !isFormattedNewValueNotAValidNumber &&
            formattedNewValue.match(/./)?.length
          ) {
            finalEventValue = String(formattedNewValue);
          } else if (isFormattedNewValueNotAValidNumber) {
            // IF the parsed number is not valid, we revert back to the valid value
            finalEventValue = value as string;
          }

          // IF 'shouldFormatToLocaleString' or 'maximumFractionDigits' are defined or the value is negative,
          // value can't have leading zeros. Like 0,000,123 or 010.50 or -00
          if (
            !isFormattedNewValueNotAValidNumber &&
            (shouldFormatToLocaleString ||
              maximumFractionDigits > 0 ||
              finalEventValue.includes("-"))
          ) {
            finalEventValue = removeLeadingZeros(locale, finalEventValue);
          }

          // IF maximumFractionDigits is set as 0, value can not be negative zero
          if (maximumFractionDigits === 0 && finalEventValue === "-0") {
            finalEventValue = "0";
          }

          // IF shouldFormatToLocaleString is defined, caret position should calculate according to thoudsandths separator count
          if (shouldFormatToLocaleString) {
            const thousandthsSeparatorCount = getThousandthSeparatorCount(
              finalEventValue
            );
            const prevValueThousandthsSeparatorCount = getThousandthSeparatorCount(
              value as string
            );
            const element = event.currentTarget;
            let caret = event.currentTarget.selectionStart || 0;

            if (prevValueThousandthsSeparatorCount === thousandthsSeparatorCount + 1) {
              caret -= 1;
            } else if (
              prevValueThousandthsSeparatorCount ===
              thousandthsSeparatorCount - 1
            ) {
              caret += 1;
            }

            window.requestAnimationFrame(() => {
              element.selectionStart = caret;
              element.selectionEnd = caret;
            });
          }

          event.currentTarget.value = finalEventValue;
        }
      }

      onChange(event);
    }
  }
  /* eslint-enable complexity */
);

export default Input;
