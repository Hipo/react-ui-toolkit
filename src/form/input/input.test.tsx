import React from "react";
import {render, cleanup, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import Input, {InputProps} from "./Input";
// import {testA11y} from "../../core/utils/test/testUtils";

describe("<Input />", () => {
  afterEach(cleanup);

  const defaultInputProps: InputProps = {
    testid: "input",
    name: "input",
    onChange: jest.fn()
  };

  it("should render correctly", () => {
    render(<Input {...defaultInputProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<Input {...defaultInputProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  //   it("should pass a11y test", async () => {
  //     const {container} = render(
  //       <Input {...defaultInputProps}  />
  //     );

  //     await testA11y(container);
  //   });

  it("should call onChange prop correctly", () => {
    render(<Input {...defaultInputProps} />);

    const input = document.getElementsByTagName("input")[0];

    fireEvent.change(input, {target: {value: "test"}});

    expect(defaultInputProps.onChange).toHaveBeenCalled();
  });

  it("should update value on change", () => {
    render(<Input {...defaultInputProps} />);

    const input = document.getElementsByTagName("input")[0];

    fireEvent.change(input, {target: {value: "test"}});

    expect(input.value).toEqual("test");
  });

  it("should render left and right icons correctly", () => {
    const iconContent = <p data-testid={"icon"}>{"Test"}</p>;

    const {rerender, getByTestId} = render(
      <Input leftIcon={iconContent} {...defaultInputProps} />
    );

    const leftIcon = getByTestId("icon");

    expect(getByTestId(defaultInputProps.testid!)).toContainElement(leftIcon);

    rerender(<Input rightIcon={iconContent} {...defaultInputProps} />);

    const rightIcon = getByTestId("icon");

    expect(getByTestId(defaultInputProps.testid!)).toContainElement(rightIcon);
  });

  it("isDisabled property should works correctly", () => {
    render(<Input isDisabled={true} {...defaultInputProps} />);

    const input = document.getElementsByTagName("input")[0];

    expect(input).toHaveAttribute("disabled");
    expect(input).toBeDisabled();
    expect(input).toHaveClass("input--is-disabled");
  });

  it("hasError property should works correctly", () => {
    render(<Input hasError={true} {...defaultInputProps} />);

    const input = document.getElementsByTagName("input")[0];

    expect(input).toHaveClass("input--has-error");
  });

  it("should format number input correctly", () => {
    render(
      <Input maxLength={3} type={"number"} maxFractionDigits={2} {...defaultInputProps} />
    );

    const input = document.getElementsByTagName("input")[0];

    expect(input).toHaveAttribute("type", "text");

    fireEvent.change(input, {target: {value: "1545.5478"}});

    expect(input.value).toEqual("1545.54");
  });
});
