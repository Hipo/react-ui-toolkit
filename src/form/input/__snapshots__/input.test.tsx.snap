// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should match snapshot 1`] = `
<div
  className="input-container"
  data-testid="input"
>
  <input
    autoComplete="off"
    autoCorrect="off"
    className="input"
    name="input"
    onChange={[Function]}
    type="text"
  />
</div>
`;
